File: tools/mksprites.py
#!/usr/bin/env python3
"""Write the cell sprite sheets.

    python3 tools/mksprites.py

One sheet per cell state. A sheet is 256x256: a 16x16 grid of 16x16 tiles, so a
cell's u8 UV picks one tile. Tile (0,0) is the default art; the rest is room for
multi-cell pictures, where a pane spanning a rectangle gives each cell a
different tile so the parts line up.

RGBA, where R is saturation, G is lightness and A is coverage. There is no hue:
it comes from the player at draw time, so one sheet serves every player. Edges
are hard -- no anti-aliasing, these are pixel art.

Standalone Python rather than a cargo example, because the crate embeds these
files with include_bytes! and so cannot build until they exist.
"""
import struct, zlib, pathlib

TILE, GRID = 16, 16
SIZE = TILE * GRID

# character -> (saturation, lightness, coverage)
INK = {
    " ": (0.00, 0.00, 0.00),
    ".": (0.85, 0.42, 0.35),
    "-": (0.85, 0.55, 0.75),
    "#": (0.85, 0.62, 1.00),
    "=": (0.70, 0.88, 1.00),
}

def blank():
    return [[(0, 0, 0, 0)] * SIZE for _ in range(SIZE)]

def stamp(sheet, tx, ty, rows):
    assert len(rows) == TILE, f"tile is {len(rows)} rows, not {TILE}"
    for y, row in enumerate(rows):
        assert len(row) == TILE, f"row {y} is {len(row)} wide, not {TILE}"
        for x, c in enumerate(row):
            s, l, a = INK[c]
            sheet[ty * TILE + y][tx * TILE + x] = (
                round(s * 255), round(l * 255), 0, round(a * 255))

def write(name, sheet):
    raw = b"".join(b"\x00" + b"".join(bytes(px) for px in row) for row in sheet)
    def chunk(tag, data):
        c = tag + data
        return struct.pack(">I", len(data)) + c + struct.pack(">I", zlib.crc32(c))
    png = (b"\x89PNG\r\n\x1a\n"
           + chunk(b"IHDR", struct.pack(">IIBBBBB", SIZE, SIZE, 8, 6, 0, 0, 0))
           + chunk(b"IDAT", zlib.compress(raw, 9))
           + chunk(b"IEND", b""))
    path = pathlib.Path("assets/sprites") / f"{name}.png"
    path.write_bytes(png)
    print(f"wrote {path}  {SIZE}x{SIZE}")

EMPTY = ["                "] * 16

ALIVE = [
    "                ",
    "   ##########   ",
    "  ############  ",
    " ############## ",
    " ############## ",
    " ############## ",
    " ############## ",
    " ############## ",
    " ############## ",
    " ############## ",
    " ############## ",
    " ############## ",
    " ############## ",
    "  ############  ",
    "   ##########   ",
    "                ",
]

PANE = [
    "================",
    "=..............=",
    "=..............=",
    "=..............=",
    "=..............=",
    "=..............=",
    "=..............=",
    "=..............=",
    "=..............=",
    "=..............=",
    "=..............=",
    "=..............=",
    "=..............=",
    "=..............=",
    "=..............=",
    "================",
]

PANE_OVER_ALIVE = [
    "================",
    "=.------------.=",
    "=.------------.=",
    "=.------------.=",
    "=.------------.=",
    "=.------------.=",
    "=.------------.=",
    "=.------------.=",
    "=.------------.=",
    "=.------------.=",
    "=.------------.=",
    "=.------------.=",
    "=.------------.=",
    "=.------------.=",
    "=.------------.=",
    "================",
]

# Dead and ice-free is deliberately blank: the file exists so the set is
# complete, and so there is somewhere to draw rubble later.
for name, art in [
    ("dead", EMPTY),
    ("alive", ALIVE),
    ("dead_ice", PANE),
    ("alive_ice", PANE_OVER_ALIVE),
]:
    sheet = blank()
    stamp(sheet, 0, 0, art)
    write(name, sheet)

File: examples/mksprites.rs
//! Write the sprite files. Run once; the PNGs are checked in and editable.
//!
//!     cargo run --example mksprites
//!
//! Each sprite is strictly 16x16 with hard edges — no anti-aliasing, so a cell
//! is pixel art rather than a smooth blob. RGBA, where R is saturation, G is
//! lightness and A is coverage. There is deliberately no hue: it comes from the
//! player at draw time, so one sprite serves every player.
//!
//! The art is ASCII here so a change is legible in a diff. Editing the PNG in
//! an image editor works just as well; this only regenerates the defaults.

const N: usize = 16;

/// ' ' empty, '.' faint, '-' mid, '#' solid, '=' bright edge.
fn encode(rows: [&str; N]) -> Vec<u8> {
    let mut out = vec![0u8; N * N * 4];
    for (y, row) in rows.iter().enumerate() {
        let chars: Vec<char> = row.chars().collect();
        assert_eq!(chars.len(), N, "row {y} is {} wide, not {N}", chars.len());
        for (x, &c) in chars.iter().enumerate() {
            let (sat, light, alpha) = match c {
                ' ' => (0.0, 0.0, 0.0),
                '.' => (0.85, 0.42, 0.35),
                '-' => (0.85, 0.55, 0.75),
                '#' => (0.85, 0.62, 1.00),
                '=' => (0.70, 0.88, 1.00),
                other => panic!("unknown sprite character {other:?}"),
            };
            let at = (y * N + x) * 4;
            out[at] = (sat * 255.0) as u8;
            out[at + 1] = (light * 255.0) as u8;
            out[at + 2] = 0;
            out[at + 3] = (alpha * 255.0) as u8;
        }
    }
    out
}

fn write(name: &str, rows: [&str; N]) {
    let path = format!("assets/sprites/{name}.png");
    let file = std::fs::File::create(&path).expect("create sprite");
    let mut encoder = png::Encoder::new(std::io::BufWriter::new(file), N as u32, N as u32);
    encoder.set_color(png::ColorType::Rgba);
    encoder.set_depth(png::BitDepth::Eight);
    encoder
        .write_header()
        .expect("png header")
        .write_image_data(&encode(rows))
        .expect("png data");
    println!("wrote {path}");
}

fn main() {
    // A living cell: a solid block with its corners cut, so a lone cell reads
    // as a rounded tile and a run of them reads as a wall.
    write("normal", [
        "                ",
        "   ##########   ",
        "  ############  ",
        " ############## ",
        " ############## ",
        " ############## ",
        " ############## ",
        " ############## ",
        " ############## ",
        " ############## ",
        " ############## ",
        " ############## ",
        " ############## ",
        "  ############  ",
        "   ##########   ",
        "                ",
    ]);

    // A pane: a bright frame with a faint fill, so whatever it covers still
    // reads through it.
    write("glass", [
        "================",
        "=..............=",
        "=..............=",
        "=..............=",
        "=..............=",
        "=..............=",
        "=..............=",
        "=..............=",
        "=..............=",
        "=..............=",
        "=..............=",
        "=..............=",
        "=..............=",
        "=..............=",
        "=..............=",
        "================",
    ]);
}
